### PR TITLE
Batch rows on access node for distributed COPY

### PR DESCRIPTION
**Pull request by akuzm**
Group the incoming rows into batches on access node before COPYing to
data nodes.

This gives 2x-5x speedup on various COPY queries to distributed
hypertables.

Also fix the text format passthrough, and prefer text transfer format for text input to be able to use this passthrough. It saves a lot of CPU on the access node.
Fixes https://github.com/timescale/timescaledb/issues/4761

Part of #4285

### CODE REVIEW

File: tsl/src/remote/dist_copy.c [hunk 1]
```diff
@@ -636,7 +950,8 @@ parse_next_text_row(CopyFromState cstate, List *attnums, TextCopyContext *ctx)
 	StringInfo row_data = makeStringInfo();
 	int i;
 
-	if (!NextCopyFromRawFields(cstate, &ctx->fields, &ctx->nfields))
+	bool result = NextCopyFromRawFields(cstate, &ctx->fields, &ctx->nfields);
+	if (!result)
```
Comment (mfundul) [inline tsl/src/remote/dist_copy.c:954]: Maybe undo this change?

Comment (akuzm) [reply tsl/src/remote/dist_copy.c:954]: Yep, this was for debugging I guess

File: tsl/src/remote/dist_copy.c [hunk 2]
```diff
@@ -230,53 +265,351 @@ calculate_hyperspace_point_from_fields(char **data, CopyDimensionInfo *dimension
...
+				 "the connection is expected to be in PGRES_COPY_IN status, but it has no activity (when flushing data)");
+		}
+
+		if (PQresultStatus(res) != PGRES_COPY_IN)
+		{
+			char *sqlstate = PQresultErrorField(res, PG_DIAG_SQLSTATE);
+			if (sqlstate != NULL && strcmp(sqlstate, "00000") == 0)
+			{
+				/*
+				 * An error have occurred.
```
Comment (mfundul) [inline tsl/src/remote/dist_copy.c:387]:
```suggestion
				 * An error has occurred.
```

File: tsl/src/remote/connection.c [hunk 3]
```diff
@@ -858,6 +858,14 @@ remote_connection_get_status(const TSConnection *conn)
 const char *
 remote_connection_node_name(const TSConnection *conn)
 {
+#ifndef NDEBUG
```
Comment (mfundul) [inline tsl/src/remote/connection.c:861]: Why don't we use `#ifdef DEBUG` like the rest of the code? I guess at this point in time `#ifdef DEBUG` should be identical to `#ifndef NDEBUG`. Is this always going to be the case? Just wondering about consistency.

Comment (akuzm) [reply tsl/src/remote/connection.c:861]: `NDEBUG` is in the C standard, and I'm not sure where the `DEBUG` comes from, so just using the more standard thing.

Comment (mfundul) [reply tsl/src/remote/connection.c:861]: We explicitly introduced DEBUG in our codebase I believe, probably many years ago. `NDEBUG` is standard when it exists, but we don't explicitly define it. Currently, `NDEBUG` is generated by CMake source code. This could be different if we change a build system, or if we use a different compiler platform or something.

In any case, I think we should either move everything to `DEBUG` or everything to `NDEBUG`. This is probably out of scope for this PR.

Comment (akuzm) [reply tsl/src/remote/connection.c:861]: > NDEBUG is standard when it exists, but we don't explicitly define it.

Well, that's the point, it's mandated by the C standard to be defined by the standard headers or tools, so it's the same e.g. on different platforms, and we don't have to define it ourselves.

Comment (mfundul) [reply tsl/src/remote/connection.c:861]: It's not mandated to be defined as far as I know. It's just mandated to skip `assert.h` if it exists (in the C standard). There are various discussions about it in the internetz, it's not very clear cut. In any case, I am in favour of using `NDEBUG` everywhere replacing `DEBUG`, but this is probably not the PR to do this.

**Review by mkindahl — COMMENTED** [review]
I see no obvious problems with the code, but given that it is a quite drastic change on how nodes are copied to the data nodes, I wonder how good the test coverage is.

Looking through the `remote_copy`, I see that we use a replication factor of 2 (good), but it might be a good idea to test with a few other replication factors as well (in particular 1, but that might already be tested elsewhere) and also testing a few more corner-cases such as copying empty rows (to a table with no columns) and copying no rows at all (possibly a few more as well).

File: tsl/src/remote/dist_copy.c [hunk 4]
```diff
@@ -54,13 +60,36 @@ typedef struct CopyDimensionInfo
 	int32 atttypmod;
 } CopyDimensionInfo;
 
+typedef struct DataNodeConnection
+{
+	TSConnectionId id;
+	TSConnection *connection;
```
Comment (pmwkaa) [inline tsl/src/remote/dist_copy.c:66]: Curious why we don't we have `TSConnectionId` field somewhere inside the `TSConnection` object, maybe it would make more sense then creating a new connection object?

File: tsl/src/remote/dist_copy.c [hunk 5]
```diff
@@ -15,31 +17,35 @@
...
-	int32 chunk_id;
-	List *connections;
-} ChunkConnectionList;
+#define MAX_BATCH_ROWS 1024
+
+/*
+ * Maximum bytes of COPY data in batch. This is also the default size of the
+ * output copy data buffer.
+ */
+#define MAX_BATCH_BYTES 10 * 1024 * 1024
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/remote/dist_copy.c:48]: IDK, postgres uses a define for this as well (MAX_BUFFERED_BYTES, copyfrom.c). I don't see an obvious need to make this configurable. Agree that it might be good in principle for advanced users, but on the other hands we shouldn't have too much options or the product becomes hard to use.

File: tsl/src/remote/dist_copy.c [hunk 6]
```diff
@@ -230,53 +265,351 @@ calculate_hyperspace_point_from_fields(char **data, CopyDimensionInfo *dimension
...
+			connection = remote_dist_txn_get_connection(required_id, REMOTE_TXN_NO_PREP_STMT);
+
+			DataNodeConnection *entry = palloc(sizeof(DataNodeConnection));
+			entry->connection = connection;
+			entry->id = required_id;
+
+			state->data_node_connections = lappend(state->data_node_connections, entry);
+			result = lappend(result, connection);
+		}
+
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/remote/dist_copy.c]: It turned out that this and surrounding code is also duplicated in another place, so I extracted it all into a function.

File: tsl/src/remote/dist_copy.c [hunk 7]
```diff
@@ -797,36 +1115,230 @@ send_copy_data(StringInfo row_data, const List *connections)
...
+			if (lc2 == NULL)
+			{
+				/* No insert state for this data node yet. Create it. */
+				data_node_rows = palloc(sizeof(DataNodeRows));
+				data_node_rows->server_oid = chunk_data_node->foreign_server_oid;
+				data_node_rows->rows_total = 0;
+				data_node_rows->row_indices = palloc(sizeof(int) * context->current_batch_rows);
+				data_nodes = lappend(data_nodes, data_node_rows);
+			}
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/remote/dist_copy.c]: It's a matter of taste, but to me a wall of code is more readable than a sequence of function calls. If it were used in several places, I'd extract it, but the entire DataNodeRows stuff is private to this function. Postgres follows the same approach -- not extracting functions just because the parent function is too long.

File: tsl/src/remote/dist_copy.c (hunk 4 above)
Comment (akuzm) [reply tsl/src/remote/dist_copy.c:66]: Probably... I'm just trying to avoid large refactorings, this PR is big enough already.

**Comment by akuzm** [comment]
> Looking through the remote_copy, I see that we use a replication factor of 2 (good), but it might be a good idea to test with a few other replication factors as well (in particular 1, but that might already be tested elsewhere) and also testing a few more corner-cases such as copying empty rows (to a table with no columns) and copying no rows at all (possibly a few more as well).

I added some tests with different replication factors and numbers of rows to `dist_copy_long`. Not sure how to test it w/o columns, is it possible to create a hypertable w/o a time column?

File: tsl/src/remote/dist_copy.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (mkindahl) [reply tsl/src/remote/dist_copy.c]: Mostly concerned about code duplication.

For shorter swaths of code I do not think moving it to a function makes a difference, for anything that you can't immediately grasp with a single look, or which requires a lot of scrolling, I usually factor it out.

**Review by mkindahl — APPROVED** [review]

**Comment by akuzm** [comment]
Turns out the text COPY passthrough is just totally broken :weary: 

https://github.com/timescale/timescaledb/issues/4761

File: test/src/test_utils.c [hunk 8]
```diff
@@ -125,19 +126,34 @@ ts_debug_shippable_fatal_after_n_rows(PG_FUNCTION_ARGS)
...
- * Use ERROR, not FATAL, because PG versions < 14 are unable to report a FATAL
- * error to the access node before closing the connection, so the test results
- * would be different.
+ * chosen prime or configured) number of rows.
  */
-#define ARBITRARY_PRIME_NUMBER 7103
+static void
+broken_sendrecv_throw()
+{
+	int throw_after = 7103; /* an arbitrary prime */
```
Comment (pmwkaa) [inline test/src/test_utils.c:134]: nit: just a thought, that could also use random number generator with seed to have a broader coverage  and deterministic output

**Review by pmwkaa — APPROVED** [review]
Looked thought it and didn't find any explicit issues. This is a big PR and a bit difficult to follow because of that, not sure if it was possible to split it into smaller pieces, but I don't want you to spend more time on it.

File: tsl/src/remote/dist_copy.c [hunk 9]
```diff
@@ -797,36 +1286,173 @@ send_copy_data(StringInfo row_data, const List *connections)
...
+
+			if (lc2 == NULL)
+			{
+				/* No insert state for this data node yet. Create it. */
+				data_node_rows = palloc(sizeof(DataNodeRows));
+				data_node_rows->connection = NULL;
+				data_node_rows->server_oid = chunk_data_node->foreign_server_oid;
+				data_node_rows->rows_total = 0;
+				data_node_rows->row_indices = palloc(sizeof(int) * context->batch_row_count);
```
Comment (erimatnor) [inline tsl/src/remote/dist_copy.c:1373]: I assume that this is oversizing this array since not all tuples in the batch are sent to all nodes? Just checking, to ensure that we need not re-size this later or are undersizing, for example.

File: tsl/src/remote/dist_copy.c [hunk 10]
```diff
@@ -797,36 +1286,173 @@ send_copy_data(StringInfo row_data, const List *connections)
...
+		}
 
-	return send_copy_data(context->row_data, connections);
+		/*
+		 * We don't have to specially flush the data here, because the flush is
+		 * attempted after finishing each protocol message (pqPutMsgEnd()).
+		 */
+	}
+
+	return true;
```
Comment (erimatnor) [inline tsl/src/remote/dist_copy.c:1442]: I guess this return value is pointless and we can remove it, instead returning void?

File: tsl/src/remote/dist_copy.c [hunk 11]
```diff
@@ -840,34 +1466,59 @@ remote_distributed_copy(const CopyStmt *stmt, CopyChunkState *ccstate, List *att
...
+			{
+				break;
+			}
 		}
 	}
 	PG_CATCH();
 	{
 		/* If we hit an error, make sure we end our in-progress COPYs */
-		remote_copy_end(context);
+		end_copy_on_failure(&context->connection_state);
```
Comment (erimatnor) [inline tsl/src/remote/dist_copy.c:1515]: suggest naming this similar/analagous to the success function. Maybe also taking just `context`?

```suggestion
		remote_copy_end_on_failure(context);
```

**Review by erimatnor — APPROVED** [review]
LGTM. A few nits, suggestions, and questions here and there.

File: tsl/src/remote/dist_copy.c [hunk 12]
```diff
@@ -230,62 +265,393 @@ calculate_hyperspace_point_from_fields(char **data, CopyDimensionInfo *dimension
...
+			break;
+		}
+	}
+
+	if (connection == NULL)
+	{
+		/*
+		 * Did not find a cached connection, create a new one and cache it.
+		 */
+		connection = remote_dist_txn_get_connection(required_id, REMOTE_TXN_NO_PREP_STMT);
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/remote/dist_copy.c:294]: The dist_copy or data_node_copy assumes it's the single user of the connection cache, basically. It takes care of exiting the COPY subprotocol if it needs to create new chunks and so on. I think we don't want other connections to be created, besides the ones we use for COPY.

File: test/src/test_utils.c (hunk 8 above)
Comment (akuzm) [reply test/src/test_utils.c:134]: Yep, could be nice, although we don't have a way to set random seed for the entire test run now. And we'll have to mask the particular numbers from the output somehow.

File: tsl/src/remote/dist_copy.c [hunk 13]
```diff
@@ -630,27 +1042,101 @@ remote_copy_get_copycmd(RemoteCopyContext *context)
 	return context->connection_state.outgoing_copy_cmd;
 }
 
+/*
+ * Functions for escaping values for Postgres text format.
+ * See CopyAttributeOutText.
+ */
+static bool
+is_special_character(char c)
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/remote/dist_copy.c:1050]: Do we ever add const for PODs passed by value? For big functions I actually think this is a good idea, but it looks kinda weird.

File: tsl/src/remote/dist_copy.c (hunk 9 above)
Comment (akuzm) [reply tsl/src/remote/dist_copy.c:1384]: Yep, added a comment.

File: tsl/src/remote/dist_copy.c [hunk 14]
```diff
@@ -630,27 +1042,101 @@ remote_copy_get_copycmd(RemoteCopyContext *context)
...
+		ereport(ERROR,
+				(errcode(ERRCODE_BAD_COPY_FILE_FORMAT),
+				 errmsg("the number of columns doesn't match")));
+	}
+
+	/*
+	 * Format the raw unquoted unescaped fields into the Postgres text format.
+	 * We only have to escape the special characters, mirroring what
+	 * CopyAttributeOutText does. We irresponsibly assume here that the server
+	 * encodings are the same and they embed ASCII, like UTF8.
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/remote/dist_copy.c:1095]: We assume here that the database encodings of the access node and the data nodes, and the client encoding that the access node uses to connect to the data nodes, are the same. The client encoding can be anything we wish, the raw fields are converted to access node encoding on the access node. I'll reword the comment and a test with SHIFT-JIS or something.

File: tsl/src/remote/dist_copy.c [hunk 15]
```diff
@@ -630,27 +1042,101 @@ remote_copy_get_copycmd(RemoteCopyContext *context)
...
+		const char * restrict src = fields[field];
+		const int input_len = (src != NULL) ? strlen(src) : 1;
+		/* Each character replaced with two, plus separator, terminating newline
+		 * and zero. */
+		const int additional_len = (input_len * 2 + 3);
+		if (row_data->maxlen < len + additional_len)
+		{
+			row_data->len = len;
+			enlargeStringInfo(row_data, additional_len);
+			output = row_data->data;
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/remote/dist_copy.c:1119]: Added a test.

File: tsl/src/remote/dist_copy.c [hunk 16]
```diff
@@ -630,27 +1042,101 @@ remote_copy_get_copycmd(RemoteCopyContext *context)
...
+	 * encodings are the same and they embed ASCII, like UTF8.
+	 */
+	const int nfields = ctx->nfields;
+	char ** restrict fields = ctx->fields;
+	char * restrict output = row_data->data;
+	int len = row_data->len;
+	for (int field = 0; field < nfields; field++)
+	{
+		const char * restrict src = fields[field];
+		const int input_len = (src != NULL) ? strlen(src) : 1;
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/remote/dist_copy.c:1115]: 1 is length of a Null value that is written as '\N' -- same size as one byte escaped. I added a comment.

File: tsl/src/remote/dist_copy.c [hunk 17]
```diff
@@ -630,27 +1042,101 @@ remote_copy_get_copycmd(RemoteCopyContext *context)
...
 
-	for (i = 0; i < ctx->nfields - 1; ++i)
-		appendStringInfo(row_data,
-						 "%s%c",
-						 ctx->fields[i] ? ctx->fields[i] : ctx->null_string,
-						 ctx->delimiter);
+		/* Add separator. */
+		if (field > 0)
+		{
+			output[len++]= '\t';
```
[the remark this answers is not in the document]
Comment (akuzm) [reply tsl/src/remote/dist_copy.c]: We have to do the escaping and the corresponding expansion manually anyway. If we insert a function call from another .so here, the compiler has less chances to optimize the loop, so I inlined it manually, given that it's not very big. This is an especially performance-critical place so I think it makes sense. Although I don't have much hopes on getting reasonably efficient code here -- the input is zero-terminated, not aligned and not padded, so we're down to byte-by-byte processing.
